components/datetime/src/provider/date_time.rs

```diff
@@ -8,102 +8,141 @@ use crate::fields;
         match options {
-            DateTimeFormatOptions::Length(bag) => self.get_pattern_for_length_bag(bag),
-            DateTimeFormatOptions::Components(bag) => self.get_pattern_for_components_bag(bag),
+            DateTimeFormatOptions::Length(bag) => self.pattern_for_length_bag(bag),
+            DateTimeFormatOptions::Components(bag) => self.pattern_for_components_bag(bag),
         }
     }
 
-    fn get_pattern_for_components_bag(
-        &self,
-        components: &components::Bag,
-    ) -> Result<Option<Pattern>> {
-        // Not all skeletons are currently supported.
-        let requested_fields = components.to_vec_fields();
-        Ok(
-            match skeleton::create_best_pattern_for_fields(
-                &self.datetime.skeletons,
-                &self.datetime.length_patterns,
-                &requested_fields,
-                components,
-                false, // Prefer the requested fields over the matched pattern.
-            ) {
-                skeleton::BestSkeleton::AllFieldsMatch(pattern)
-                | skeleton::BestSkeleton::MissingOrExtraFields(pattern) => Some(pattern.0),
-                skeleton::BestSkeleton::NoMatch => None,
-            },
-        )
+    fn date_patterns(&mut self) -> Result<&DataPayload<'data, DatePatternsV1Marker>> {
+        if let Some(ref date_patterns) = self.date_patterns {
+            Ok(date_patterns)
+        } else {
+            let patterns_data: DataPayload<'data, DatePatternsV1Marker> = self
+                .data_provider
+                .load_payload(&DataRequest {
+                    resource_path: ResourcePath {
+                        key: provider::key::GREGORY_DATE_PATTERNS_V1,
+                        options: ResourceOptions {
+                            variant: None,
+                            langid: Some(self.locale.clone().into()),
+                        },
+                    },
+                })?
+                .take_payload()?;
+            Ok(self.date_patterns.insert(patterns_data))
+        }
+    }
+
+    fn date_skeletons(
+        &mut self,
+    ) -> Result<(
+        &DataPayload<'data, DatePatternsV1Marker>,
+        &DataPayload<'data, DateSkeletonPatternsV1Marker>,
+    )> {
+        let patterns = if let Some(ref date_patterns) = self.date_patterns {
+            date_patterns
+        } else {
+            let patterns_data: DataPayload<'data, DatePatternsV1Marker> = self
+                .data_provider
+                .load_payload(&DataRequest {
+                    resource_path: ResourcePath {
+                        key: provider::key::GREGORY_DATE_PATTERNS_V1,
+                        options: ResourceOptions {
+                            variant: None,
+                            langid: Some(self.locale.clone().into()),
+                        },
+                    },
+                })?
+                .take_payload()?;
+            self.date_patterns.insert(patterns_data)
+        };
```

Comment:
This is one thing I don't love about this patch - it duplicates `skeletons()` but because of mutable borrow checking I don't have an easy way to have `patterns()` separately from `skeletons()` without playing with `RefCell`.

I think it's an ok tradeoff for now, and we can polish that when we have the DTF more settled after 0.4.

Reply:
I was able to fix this by separating loading logic into fields which satisfies the borrow checker.